GoogleForCreators/web-stories-wp
# Design System: Focus first focusable item in the context menu

## Context

Big thanks to @littlemilkstudio for helping to solve this.

This used to exist but **it didn't work in Firefox**. https://github.com/google/web-stories-wp/pull/6670/files#diff-4e64e6be1f54bcbbded9bc4053ade1162fd6fb817b8f6c3427f1aa07f888058bL166-L193. We removed it.

This PR is another attempt at adding in this functionality.

---

**THE BUG (which exists in main):**

The bug happened in firefox when:
- An item in the context menu is focused
- Click a button in the menu (like `delete story`). The event doesn't fire.
    - Note: clicking a link still works

So when we autofocused an item it would show this bug. The bug wasn't solved before, it was just avoided.

## Summary

- Focuses the first item when the context menu is opened with keyboard
- Shows box shadow (focus) when the animated context menu is open

## Relevant Technical Choices

n/a

## To-do

none

## User-facing changes

||Before|After|
|--|--|--|
|Auto focus first item|![not-focused](https://user-images.githubusercontent.com/22185279/113458672-47ae0680-93d0-11eb-859a-9633eec9c53c.gif)|![focused](https://user-images.githubusercontent.com/22185279/113458673-48df3380-93d0-11eb-8478-a906a6083a15.gif)|
|Bug fixed - clicking buttons fires actions|![broken](https://user-images.githubusercontent.com/22185279/113458432-75468000-93cf-11eb-9501-eddf0b148bb9.gif)|![fixed](https://user-images.githubusercontent.com/22185279/113458405-5b0ca200-93cf-11eb-8ece-dea25cd17710.gif)|
|Focus border shows on edges|![image](https://user-images.githubusercontent.com/22185279/113458522-c5254700-93cf-11eb-9020-b28f7554f757.png)|![image](https://user-images.githubusercontent.com/22185279/113458514-bc347580-93cf-11eb-8a0e-adecc6990bbc.png)|

## Testing Instructions

To test automatic focus of first focusable item: (all browsers)
1. Use keyboard to get to context menu button
2. Press `Space` or `Enter` to open the context menu
3. See that the first item is focused

To test bug that was fixed: **MAKE SURE TO CHECK IN FIREFOX, THAT'S WHERE THE BUG WAS BEFORE**
1. Go to dashboard
2. Click the context menu button to open the context menu for a story
3. `Press Tab` to focus an element in the menu
4. Click the `Delete story` button. A modal should show.

### QA






This PR can be tested by following these steps:

To test automatic focus of first focusable item: (all browsers)
1. Use keyboard to get to context menu button
2. Press `Space` or `Enter` to open the context menu
3. See that the first item is focused

To test bug that was fixed: **MAKE SURE TO CHECK IN FIREFOX, THAT'S WHERE THE BUG WAS BEFORE**
1. Go to dashboard
2. Click the context menu button to open the context menu for a story
3. `Press Tab` to focus an element in the menu
4. Click the `Delete story` button. A modal should show.
### UAT






## Reviews

### Does this PR have a security-related impact?

no

### Does this PR change what data or activity we track or use?

no

### Does this PR have a legal-related impact?

no

## Checklist




---



Fixes #6679
